core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -225,21 +225,25 @@ sealed abstract class IO[E, A] { self =>
    * }}}
    */
   final def bracket[B](release: A => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
-    new IO.Bracket(this, (_: ExitResult[E, B], a: A) => release(a), use)
+    IO.bracket(this) {
+      case ExitResult.Completed(a)  => release(a)
+      case ExitResult.Failed(_)     => IO.unit
+      case ExitResult.Terminated(t) => IO.terminate(t)
+    }(use)
 
   /**
    * A more powerful version of `bracket` that provides information on whether
    * or not `use` succeeded to the release action.
    */
-  final def bracket0[B](release: (ExitResult[E, B], A) => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
-    new IO.Bracket(this, release, use)
+  // final def bracket0[B](release: (ExitResult[E, B], A) => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
```

Comment:
It should be straightforward to modify `IO.bracket` to pass along the `A` from acquire. This will let you keep `bracket0`, `bracketOnError`, and `onError`.